test/hotspot/jtreg/runtime/ErrorHandling/StackWalkNativeToJava.java

```diff
@@ -0,0 +1,129 @@
+        List<String> commands = new ArrayList<>();
+        commands.add("-Xbootclasspath/a:.");
+        commands.add("-XX:-CreateCoredumpOnCrash");
+        commands.add("-XX:+UnlockDiagnosticVMOptions");
+        commands.add("-XX:AbortVMOnException=java.lang.IllegalMonitorStateException");
+        commands.add("-XX:+ErrorFileToStdout");
+        commands.addAll(Arrays.asList(extraFlags));
+        commands.add("StackWalkNativeToJava$TestNativeToJavaNative");
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(commands);
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldNotContain("java.lang.RuntimeException: Reached statement after obj.wait()");
+        output.shouldNotContain("[error occurred during error reporting (printing native stack");
+        String[] res = output.getOutput().split("StackWalkNativeToJava\\$TestNativeToJavaNative\\.callNativeMethod\\(\\)V");
+        assertTrue(res.length - 1 == 2, res.length - 1);
+        output.shouldNotHaveExitValue(0);
+    }
+
+    public static class TestNativeToJavaNative {
+        public static void main(String[] args) throws Exception {
+            TestNativeToJavaNative test = new TestNativeToJavaNative();
+            test.callNativeMethod();
+        }
+
+        public void callNativeMethod() throws Exception {
+            Object obj = new Object();
+            obj.wait();
+            throw new RuntimeException("Reached statement after obj.wait()");
+        }
+    }
+
+    public static void testStackWalkNativeToJava(String... extraFlags) throws Exception {
+        List<String> commands = new ArrayList<>();
+        commands.add("-Xbootclasspath/a:.");
+        commands.add("-XX:-CreateCoredumpOnCrash");
+        commands.add("-XX:+UnlockDiagnosticVMOptions");
+        commands.add("-XX:DiagnoseSyncOnValueBasedClasses=1");
+        commands.add("-XX:+ErrorFileToStdout");
+        commands.addAll(Arrays.asList(extraFlags));
+        commands.add("StackWalkNativeToJava$TestNativeToJava");
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(commands);
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldNotContain("java.lang.RuntimeException: Reached statement after synchronized");
+        output.shouldNotContain("[error occurred during error reporting (printing native stack");
+        String[] res = output.getOutput().split("StackWalkNativeToJava\\$TestNativeToJava\\.callVMMethod\\(\\)V");
+        assertTrue(res.length - 1 == 2, res.length - 1);
+        output.shouldNotHaveExitValue(0);
+    }
+
+    public static class TestNativeToJava {
+        static Integer counter = 0;
+
+        public static void main(String[] args) throws Exception {
+            TestNativeToJava test = new TestNativeToJava();
+            test.callVMMethod();
+        }
+
+        public void callVMMethod() throws Exception {
```

Comment:
Again a comment outlining how you expect this to abort the VM would be good.

Reply:
Done.